src/main/java/ch/jalu/configme/beanmapper/instantiation/BeanInstantiationServiceImpl.java

```diff
@@ -0,0 +1,81 @@
+package ch.jalu.configme.beanmapper.instantiation;
+
+import ch.jalu.configme.beanmapper.propertydescription.BeanDescriptionFactory;
+import ch.jalu.configme.beanmapper.propertydescription.BeanDescriptionFactoryImpl;
+import ch.jalu.configme.beanmapper.propertydescription.BeanFieldPropertyDescription;
+import ch.jalu.configme.internal.record.RecordComponent;
+import ch.jalu.configme.internal.record.RecordInspector;
+import ch.jalu.configme.internal.record.ReflectionHelper;
+import org.jetbrains.annotations.NotNull;
+
+import java.lang.reflect.Constructor;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class BeanInstantiationServiceImpl implements BeanInstantiationService {
+
+    private final RecordInspector recordInspector;
+    private final BeanDescriptionFactory beanDescriptionFactory;
+    private final Map<Class<?>, BeanInstantiation> cachedInstantiationByType = new ConcurrentHashMap<>();
+
+    public BeanInstantiationServiceImpl() {
+        this.recordInspector = new RecordInspector(new ReflectionHelper());
+        this.beanDescriptionFactory = new BeanDescriptionFactoryImpl();
+    }
+
+    public BeanInstantiationServiceImpl(@NotNull RecordInspector recordInspector,
+                                        @NotNull BeanDescriptionFactory beanDescriptionFactory) {
+        this.recordInspector = recordInspector;
+        this.beanDescriptionFactory = beanDescriptionFactory;
+    }
+
+    @Override
+    public @NotNull Optional<BeanInstantiation> findInstantiation(@NotNull Class<?> clazz) {
+        BeanInstantiation cachedInstantiation = cachedInstantiationByType.get(clazz);
+        if (cachedInstantiation != null) {
+            return Optional.of(cachedInstantiation);
+        }
+
+        if (recordInspector.isRecord(clazz)) {
+            RecordComponent[] recordComponents = recordInspector.getRecordComponents(clazz);
+            List<BeanFieldPropertyDescription> properties =
+                beanDescriptionFactory.createRecordProperties(clazz, recordComponents);
+
+            BeanRecordInstantiation recordInstantiation = new BeanRecordInstantiation(clazz, properties);
+            cachedInstantiationByType.put(clazz, recordInstantiation);
+            return Optional.of(recordInstantiation);
+        }
+
+        Optional<Constructor<?>> zeroArgConstructor = tryFindConstructor(clazz);
```

Comment:
Current impl assumes the constructor is public; not sure that should be a requirement, resp. if so, let's make it more explicit and test more aggressively